pulsar-common/src/main/java/org/apache/pulsar/common/util/AsyncTokenBucket.java

```diff
@@ -0,0 +1,483 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.pulsar.common.util;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLongFieldUpdater;
+import java.util.concurrent.atomic.LongAdder;
+import java.util.function.LongSupplier;
+
+/**
+ * An asynchronous token bucket algorithm implementation that is optimized for performance with highly concurrent
+ * use. CAS (compare-and-swap) operations are used and multiple levels of CAS fields are used to minimize contention
+ * when using CAS fields. The {@link LongAdder} class is used in the hot path to hold the sum of consumed tokens.
+ * It is eventually consistent, meaning that the tokens are not updated on every call to the "consumeTokens" method.
+ * <p>Main usage flow:
+ * 1. tokens are consumed by calling the "consumeTokens" method.
+ * 2. the "calculateThrottlingDuration" method is called to calculate the duration of a possible needed pause when the
+ * tokens
+ * are fully consumed.
```

Comment:
- It would be more clear if tell user how to determine tokens are whether fully consumed.
- Is it possible to tell is there any tokens available after `consumeTokens` method exit.

Reply:
> It would be clearer if we could inform the user about how to determine whether the tokens are fully consumed.

thanks for the feedback. I'll try to clarify this by making a further refactoring. 
In the current design, tokens are eventually consistent, so it would be useful to have this part of the naming.
For the asynchronous rate limiting use cases in Pulsar, there isn't really a need to have a consistent view of tokens. The token value can also be negative and that is important for correct behavior of the rate limiter.

Reply:
@Shawyeok PTAL, I have clarified the Javadoc. I added a separate method consumeTokensAndCheckIfContainsTokens to handle the main usage flow in PublisherLimiterImpl. The docs contain explanations about the eventual consistent nature.

Reply:
@Shawyeok PTAL. I have also updated the [PIP-322 document](https://github.com/lhotari/pulsar/blob/lh-ratelimiter-refactoring-pip/pip/pip-322.md) with better explanations.